8Knot/pages/codebase/visualizations/contribution_file_heatmap.py

```diff
@@ -0,0 +1,421 @@
+    directories = df["file_path"].str.rsplit("/", n=1).str[0].tolist()
+    directories = list(set(directories))
+
+    # get all of the file names to filter out of the directory set
+    top_level_files = df["file_name"][df[1].isnull()].tolist()
+    directories = [f for f in directories if f not in top_level_files]
+
+    # sort alphabetically
+    directories = sorted(directories)
+    print(directories)
+
+    # add top level directory to the list of directories
+    directories.insert(0, "Top Level Directory")
+
+    return directories, "Top Level Directory"
+
+
+# callback for contributor file heatmap graph
+@callback(
+    Output(f"{PAGE}-{VIZ_ID}", "figure"),
+    [
+        Input(f"repo-{PAGE}-{VIZ_ID}", "value"),
+        Input(f"directory-{PAGE}-{VIZ_ID}", "value"),
+        Input(f"graph-view-{PAGE}-{VIZ_ID}", "value"),
+    ],
+    background=True,
+)
+def cntrb_file_heatmap_graph(repo_id, directory, graph_view):
+    # wait for data to asynchronously download and become available.
+    cache = cm()
+
+    # execute queries for the 3 needed dfs
+    df_file = cache.grabm(func=rfq, repos=[repo_id])
+    while df_file is None:
+        time.sleep(1.0)
+        df_file = cache.grabm(func=rfq, repos=[repo_id])
+
+    df_file_pr = cache.grabm(func=prfq, repos=[repo_id])
+    while df_file_pr is None:
+        time.sleep(1.0)
+        df_file_pr = cache.grabm(func=prfq, repos=[repo_id])
+
+    df_pr = cache.grabm(func=prq, repos=[repo_id])
+    while df_pr is None:
+        time.sleep(1.0)
+        df_pr = cache.grabm(func=prq, repos=[repo_id])
+
+    start = time.perf_counter()
+    logging.warning(f"{VIZ_ID}- START")
+
+    # test if there is data
+    if df_file.empty or df_file_pr.empty or df_pr.empty:
+        logging.warning(f"{VIZ_ID} - NO DATA AVAILABLE")
+        return nodata_graph
+
+    print(directory)
+
+    # function for all data pre processing
+    df = process_data(df_file, df_file_pr, df_pr, directory, graph_view)
+
+    # if there are no pull request on a directory plot no data graph
+    if df.empty:
+        return nodata_graph
+
+    fig = create_figure(df, graph_view)
+
+    logging.warning(f"{VIZ_ID} - END - {time.perf_counter() - start}")
+    return fig
+
+
+def process_data(df_file: pd.DataFrame, df_file_pr: pd.DataFrame, df_pr: pd.DataFrame, directory, graph_view):
+
+    # strings to hold the values for each column (always the same for every row of this query)
+    repo_name = df_file["repo_name"].iloc[0]
+    repo_path = df_file["repo_path"].iloc[0]
+    repo_id = str(df_file["id"].iloc[0])
+
+    # pattern found in each file path, used to slice to get only the root file path
+    path_slice = repo_id + "-" + repo_path + "/" + repo_name + "/"
+    df_file["file_path"] = df_file["file_path"].str.rsplit(path_slice, n=1).str[1]
+
+    # drop columns not in the most recent collection
+    df_file = df_file[df_file["rl_analysis_date"] == df_file["rl_analysis_date"].max()]
```

Comment:
since we're doing this every time, maybe we should lift this into the sql query so less data has to go from the server to cache. In the future, we could let the user decide what date to see the activity up to, e.g. "what was the state of the repo in august of 2008?"

Reply:
So unfortunately the  rl_analysis_date is the date augur collected it so it will only be able to go back to the first time the repo was collected on

Reply:
okay, then this should be moved into SQL imo. the current SQL is moving more data than it needs to, a subquery would give the max analysis date, and then you can condition on the rl_analysis_date matching that.

Reply:
oop you already did it!